ensime_shared/ensime.py

```diff
@@ -300,3 +362,13 @@ def fun_en_complete_func(self, client, findstart_and_base, base=None):
             # Invoked by vim
             findstart = findstart_and_base
         return client.complete_func(findstart, base)
+
+    def watchdog_client_server_connection(self, client):
+        """Run periodically to check and maintain client, server, and cxn health."""
+        if not client.connected and client.connection_attempts < 10:
+            config = client.config
+            server = self.server(config)
+            if not server.isrunning():
+                self.start_server(config)
+            client.connect(server)
+            client.connection_attempts += 1
```

Comment:
I think experimentation won't hurt and if it turns out to be unnecessary I am more than welcome to remove it. However, there is a tradeoff here between handlers and responsiveness. If me remove `CursorHold`, `ensime-vim` could turn less responsive and responses may be processed with delays. I hope this is not the case though, I would very much prefer to use only one -- but let's keep in mind that if we're browsing through a file and waiting for some reaction of the plugin, this won't happen until the cursor is on hold for a few secs.